### ISSUE TITLE
PME multiprocessing problem on CPU platform

### ISSUE THREAD

**Issue by Olllom** [openmm/openmm#2602]
The PME implementation on the CPU platform seems to corrupt memory.
The example below hangs after the second energy call.
There are several ways to avoid the error:

1. don't use PME
2. use the Reference platform
3. don't compute energies
4. compute energies only in spawned processes
5. don't compute energy on a spawned process after the parent process has already done an energy calculation

see comments in the example.
 

```
from simtk.openmm import LangevinIntegrator, Context, Platform, NonbondedForce
from openmmtools.testsystems import WaterBox
import multiprocessing as mp

def run():
    testsystem = WaterBox()
    
    for i, force in reversed(list(enumerate(testsystem.system.getForces()))):
        if isinstance(force, NonbondedForce):
            print(force.getNonbondedMethod())
            force.setNonbondedMethod(NonbondedForce.PME)
            # Everything works fine if we comment in either of the following lines
            # force.setNonbondedMethod(NonbondedForce.NoCutoff)
            # force.setNonbondedMethod(NonbondedForce.CutoffPeriodic)
            # force.setNonbondedMethod(NonbondedForce.Ewald)
        
    # platform = Platform.getPlatformByName("CPU")
    # Everything works fine if we comment in only the following line
    platform = Platform.getPlatformByName("Reference")
    
    context = Context(
        testsystem.system, 
        LangevinIntegrator(300,1,0.001), 
        platform
    )
    context.setPositions(testsystem.positions)
    
    # Everything works fine if we set getEnergy=False
    context.getState(getPositions=True, getEnergy=True)      
    print("done")

    
# Everything works fine if we comment out only the following lines:
process = mp.Process(target=run)
process.start()
process.join()

# Everything works fine if we comment out only the following line:
run()

# Everything works fine if we comment out only the following lines:
process = mp.Process(target=run)
process.start()
process.join()
```

**Comment by RagnarB83** [comment]
Hi, I just came accross this problem as well: Running OpenMM in parallel using Python multiprocessing causes hangs on the CPU platform. No issue on the OpenCL or Reference platform.

Has there been any update on this?

**Comment by peastman** [comment]
When I run the script above, I get an exception:

```
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

Following the instructions at https://stackoverflow.com/questions/18204782/runtimeerror-on-windows-trying-python-multiprocessing, I put the creation of the subprocesses inside a `if __name__ == '__main__'` check.  Once I do that, it runs correctly.  Does that fix it for you?

**Comment by peastman** [comment]
I managed to reproduce the problem on a different computer.  It seems to be happening inside PocketFFT.  This is the line in the OpenMM code that hangs:

https://github.com/openmm/openmm/blob/3dc2f12ff20efcb609dfa5508431065d2c98d028/platforms/reference/src/SimTKReference/ReferencePME.cpp#L800
